distributed/shuffle/_shuffle_extension.py

```diff
@@ -769,11 +770,12 @@ def transition(
     ) -> None:
         if finish != "forgotten":
             return
-        if key not in self.barriers.values():
-
+        if not self._is_barrier_key(key):
+            return
+        shuffle_id = self.id_from_key(key)
+        if shuffle_id not in self.worker_for:
             return
```

Comment:
Note: I ran a couple of tests and think we do not suffer any significant performance impact here. However, I would like to ask for caution when it comes to this sort of refactoring in a transition hook. Str comparisons + replacements are not cheap, relatively speaking. Transitions are amongst the hottest for loops we have and it typically pays off to be a bit cautious here. Everything past these guards is allowed to be slow since we're only executing it for barriers but the guards themselves are evaluated for every task in a potentially very large graph.

This changes introduces an overhead of 100-200ns compared to the earlier version so I think this is fine. However, this can quickly spiral out of control if the logic on these methods become more complex or we open the transition to other allowed finished states, etc.

Reply:
Note: We could inline `_is_barrier_key` to shave off the function call overhead.

Reply:
I ran something like this

```python
adict = {f"foo-{x}": True for x in range(10)}

a_long_key = "shuffle-barrier-123566886512341"

def foo_main():
    if finish != "forgotten":
        return
    if a_long_key in adict.values():
        return
def is_barrier():
    return a_long_key.startswith("shuffle-barrier")

def foo():
    if finish != "forgotten":
        return
    if not is_barrier():
        return
    assert a_long_key.startswith("shuffle-barrier")
    key = a_long_key.replace("shuffle-barrier", "")
    if key in adict:
        return
    return
```
![image](https://user-images.githubusercontent.com/8629629/207063251-53d37e6a-f6a4-4975-9867-c6440ce49434.png)

Reply:
> Note: We could inline _is_barrier_key to shave off the function call overhead.

Don't worry about it. I merely wanted to ensure some awareness.

Reply:
Thanks for sharing, the difference in our results makes sense. I had focused on the common case where the key is forgotten but it's not a shuffle barrier, which is coincidentally faster in your example.

> I merely wanted to ensure some awareness.
Appreciated!